rustfst-ffi/src/fst.rs

```diff
@@ -0,0 +1,359 @@
+    symt: *const CSymbolTable,
+) -> RUSTFST_FFI_RESULT {
+    wrap(|| {
+        let fst = get_mut!(CFst, fst);
+        let symt = get!(CSymbolTable, symt);
+        fst.set_input_symbols(symt.clone());
+        Ok(())
+    })
+}
+
+#[no_mangle]
+pub extern "C" fn fst_unset_input_symbols(fst: *mut CFst) -> RUSTFST_FFI_RESULT {
+    wrap(|| {
+        let fst = get_mut!(CFst, fst);
+        fst.take_input_symbols();
+        Ok(())
+    })
+}
+
+#[no_mangle]
+pub extern "C" fn fst_output_symbols(
+    fst: *const CFst,
+    mut output_symt: *mut CSymbolTable,
+) -> RUSTFST_FFI_RESULT {
+    wrap(|| {
+        let fst = get!(CFst, fst);
+        fst.output_symbols()
+            .map(|it| {
+                let symt = CSymbolTable(it.clone());
+                unsafe { *output_symt = symt }
+            })
+            .unwrap_or_else(|| output_symt = std::ptr::null_mut());
+        Ok(())
+    })
+}
+
+#[no_mangle]
+pub extern "C" fn fst_set_output_symbols(
+    fst: *mut CFst,
+    symt: *const CSymbolTable,
+) -> RUSTFST_FFI_RESULT {
+    wrap(|| {
+        let fst = get_mut!(CFst, fst);
+        let symt = get!(CSymbolTable, symt);
+        fst.set_output_symbols(symt.clone());
+        Ok(())
+    })
+}
+
+#[no_mangle]
+pub extern "C" fn fst_unset_output_symbols(fst: *mut CFst) -> RUSTFST_FFI_RESULT {
+    wrap(|| {
+        let fst = get_mut!(CFst, fst);
+        fst.take_output_symbols();
+        Ok(())
+    })
+}
+
+#[no_mangle]
+pub extern "C" fn fst_set_final(
+    fst: *mut CFst,
+    state: CStateId,
+    weight: libc::c_float,
+) -> RUSTFST_FFI_RESULT {
+    wrap(|| {
+        let fst = get_mut!(CFst, fst);
+        fst.set_final(state, TropicalWeight::new(weight as f32))?;
+        Ok(())
+    })
+}
+
+#[no_mangle]
+pub extern "C" fn fst_is_final(
+    fst: *const CFst,
+    state: CStateId,
+    is_final: *mut libc::size_t,
+) -> RUSTFST_FFI_RESULT {
+    wrap(|| {
+        let fst = get!(CFst, fst);
+        let res = fst.is_final(state)?;
+        unsafe { *is_final = res as usize }
+        Ok(())
+    })
+}
+
+#[no_mangle]
+pub extern "C" fn fst_final_weight(
+    fst: *const CFst,
+    state_id: CStateId,
+    final_weight: *mut libc::c_float,
+) -> RUSTFST_FFI_RESULT {
+    wrap(|| {
+        let fst = get!(CFst, fst);
+        let res = fst
+            .final_weight(state_id)?
+            .ok_or_else(|| anyhow!("State '{:?}' is NOT final", state_id))?;
+        unsafe { *final_weight = *res.value() };
+        Ok(())
+    })
+}
+
+#[no_mangle]
+pub extern "C" fn fst_set_start(fst: *mut CFst, state: CStateId) -> RUSTFST_FFI_RESULT {
+    wrap(|| {
+        let fst = get_mut!(CFst, fst);
+        fst.set_start(state)?;
+        Ok(())
+    })
+}
+
+#[no_mangle]
+pub extern "C" fn fst_start(fst: *const CFst, state: *mut CStateId) -> RUSTFST_FFI_RESULT {
+    wrap(|| {
+        let fst = get!(CFst, fst);
+        let res = fst
+            .start()
+            .ok_or_else(|| anyhow!("FST has no start state"))?;
```

Comment:
Do we really want to raise an error here ? 
I guess we could return a null ptr when there is no start state so that in python we can retrun None.

Reply:
Yes, I'll change that !